File: 표/stack_queue_heap/택배상자.py
"""
<문제 설명>
1 ~ n번 상자까지 오름차순으로 전달
전달된 순서대로 확인 시작
order에 맞지 않는 상자는 stack에 넣음
stack에 넣은 후에 order에 맞게 실지 못하면 종료

"""

def solution(order):
    answer = 0
    stack = [] # 보조 벨트
    
    box = 1
    for target in order:
        # box가 order보다 커지면 안됨 & box가 target보다 커도 안 됨
        while box <= len(order) and box < target:
            stack.append(box)
            box += 1
        
        # 보조 벨트에서 답 있으면
        if box == target:
            answer += 1
            box += 1
        elif stack[-1] == target:
            answer += 1
            stack.pop()
        # 답 없으면    
        else: 
            return answer
            
    return answer

File: 최윤경/택배상자.js
/*
[접근]

1. 보조 벨트는 마지막에 넣은 상자부터 꺼내므로 스택(LIFO)으로 구현
2. 목표 상자 전까지의 상자는 stack에 push
3. 현재 상자 또는 stack의 top이 목표 상자면 적재
4. 둘 다 불가능하면 종료 후 적재한 개수 반환
*/

function solution(order) {
  const stack = [];
  let box = 1;
  let count = 0;

  for (const target of order) {
    while (box < target) {
      stack.push(box);
      box++;
    }

    if (box === target) {
      box++;
      count++;
    } else if (stack.length && stack[stack.length - 1] === target) {
      stack.pop();
      count++;
    } else {
      break;
    }
  }

  return count;
}
